Medical System

costno opened the pull request:
Medical system file base and minimum changes to assignGear and AI Skills

costno committed:
Medical System

Medical system file base and minimum changes to assignGear and AI Skills

costno committed:
Additional medical related AssignGear tweaks

Large medic bags, proper medic resources

costno commented:
Short term still to do:
- Vehicle loadouts with new resources
- Medswap currently overrides FAKs in server-side assignGeared vehicles and units (must implement delay)

NikkoJT commented on f/medical/fn_famRagDoll.sqf:
```diff
@@ -0,0 +1,67 @@
+
+    player setDamage _dmg;
+    player addItem "FirstAidKit"; //replace the one they won't know they lost.
+    
```
A direct unit reference should be used rather than `player` (should it be `_unit` or some other unit?)
If this runs on an AI in this state, it will cause the effects to apply to whichever player the AI is local to, which would be bad.

NikkoJT committed:
Use lazy evaluation for performance

shadow-fa commented on f/medical/fn_famInit.sqf:
```diff
@@ -0,0 +1,68 @@
+[player] remoteExec ["f_fnc_famAddDragAction", 0, player]; 
+
+// Handle damage appropriately for the system.
+_unit call f_fnc_famDamageHandler;
```
Does every player have to add this EH to every other playable unit?
Shouldnt these things only be applied to where the unit is local?

costno replied:
I definitely have a very inconsistent handling of AI currently. Playable units was used for early stages of testing so I could test on AI, but I think in the final version it makes sense to run none of this on AI.

shadow-fa replied:
My point wasnt about AI. It was about where the EH should be added and where it should be run. Does it run whereever the EH was added or just where the unit is local? Does this get executed multiple times on multiple machines whenever someone takes damage? (I don't know a lot about EHs). If we only care about the player themself then we should probably use `player` as the argument here and also make sure that it only gets added once.

costno replied:
Adjustments have been made so that everything in the init only runs when the unit is local to yourself. Which means AI will get the damage handler and be beefier on the off chance they get shot before a real human joins the slot. 

According to Chozo's damage handler, the handler _does_ run when a unit is damaged but is not local to you (not sure if this is all units or would just be units that were once local to you). But that damage handler has a quick exit if !local which will avoid running that business once the unit is no longer local.  It is my understanding that damage is handled locally otherwise, and even if the damage handler triggers for a non-local unit it would not update a remote unit unless otherwise remoteExeced?  

Not sure if what I've got will be the best way to handle it, so leaving this comment open if others have better information.

shadow-fa commented on f/medical/fn_famInit.sqf:
```diff
@@ -0,0 +1,68 @@
+_unit setVariable ["FAM_HASFAK",false]; 
+_unit setVariable ["FAM_HASBANDAGE",false]; 
+_unit getVariable ["FAM_FLAG",false];
+FAM_MEDICMOD = 0.5;
```
Instead of `FAM_*` we could think about using our (otherwise consistent?) `f_var_*` naming scheme for global variables to avoid conflicts with mission maker variables

shadow-fa commented on f/medical/fn_famWakeUp.sqf:
```diff
@@ -0,0 +1,91 @@
+    _unit removeEventHandler [_thisEvent, _thisEventHandler];
+}];
+
+// ...Or setCaptive false after 10 seconds and return senses if unit stays awake long enough.
```
```suggestion
// ...Or setCaptive false after a few seconds and return senses if unit stays awake long enough.
```

shadow-fa commented on f/medical/fn_famPassOut.sqf:
```diff
@@ -0,0 +1,117 @@
+    _unit setVariable ["FAM_wound_down_items",(assignedItems _unit select {_x == "ItemGPS" || _x == "ItemMap"})];
+    {
+        _unit unassignItem _x;
+        _unit removeItem _x;
```
no map or gps? 😰

costno replied:
Only while you are passed out! Otherwise you can leave a million markers while unable to talk. I give ‘em back!

shadow-fa commented on f/medical/fn_famRagDoll.sqf:
```diff
@@ -0,0 +1,67 @@
+
+// RAGDOLL
+
+_unit addForce [[0,20,200], [2,0,2]]; 
```
I am worried that this could actually kill someone (if being forced into some solid object or something). We could disable damage while the unit is flying around

costno replied:
It is my understanding that addForce does set allowDamage false until you animate to unconscious, along with other hidden things like setting you unconscious. But worth testing

costno commented on init.sqf:
```diff
@@ -93,9 +93,18 @@ f_script_briefing = [] spawn f_fnc_createBriefing;
+
+{
+	_x remoteExec ["f_fnc_famInit",player,player];
+} foreach playableUnits;
```
It is in this state because JIP was not working. This does not work for JIP either though, so still looking for JIP safe method. Event handler?

shadow-fa replied:
something like this then?
`player remoteExec ["f_fnc_famInit",0];` this tells eveyrone that this player exists. so each other machine would then install the necessary handlers for that new player

shadow-fa replied:
I would like to understand the current code here which reads `_x remoteExec ["f_fnc_famInit",0,player];`.
So each player does a remoteExec on everyone else for every playable unit that exists. Does this mean that it runs n^3 times for n players?
Did my previous suggestion not work?

shadow-fa replied:
I would like to understand the current code here which reads `_x remoteExec ["f_fnc_famInit",0,player];`.
So each player does a remoteExec on everyone else for every playable unit that exists. Does this mean that it runs n^3 times for n players?
Did my previous suggestion not work?

shadow-fa replied:
I don't quite understand the latest change which reads `_x remoteExec ["f_fnc_famInit",0,player];`.
If I udnerstand it correctly every player runs f_fnc_famInit on every client for every playable unit.
Does this mean that for n players it effectively does n^3 remoteExec's? And then out of these n^3 invocations we discard n^3-1 of them since the code itself checks if the unit is local?
I think what we want to do here instead is simply `_x call f_fnc_famInit`;

costno commented on f/medical/fn_famReset.sqf:
```diff
@@ -0,0 +1,20 @@
+// F3 FA Medical
+// Credits and documentation: https://github.com/folkarps/F3/wiki
+// ====================================================================================
+
```
We have some FA tools Zeus can use, are those functions? That was the intent here, obviously not connected to Zeus properly yet.

costno commented on f/medical/fn_famLoop.sqf:
```diff
@@ -0,0 +1,149 @@
+
+	if (_unit getVariable ["FAM_FORCEDOWN",false] && _unit getVariable ["FAM_CONSCIOUS",true]) then {
+
+		_unit setVariable ["FAM_CONSCIOUS",false]; 
```
This might have been left over from a previous version, it will be removed. fnc_PassOut broadcasts you are unconscious.

costno committed:
Minor language change

Co-authored-by: shadow-fa <shadow-fa@users.noreply.github.com>

costno committed:
Remove vanilla revive from description

Forgot to remove this in initial github version upload.

costno committed:
Changes from previous comments

shadow-fa commented on f/medical/fn_famLoop.sqf:
```diff
@@ -123,8 +122,7 @@ while {alive _unit && local _unit} do {
 	// LOOP RESTART
-	// Loops every second
-	sleep 1;
+	sleep 0.1;
```
0.1 seems excessive, but if it doesn't hurt the performance I guess it's okay

costno commented on f/medical/fn_famLoop.sqf:
```diff
@@ -0,0 +1,149 @@
+
+		if (_save >= _dc) then {
+
+			_unit setVariable ["FAM_CONSCIOUS",false]; 
```
Likely another carry over / redundant variable update, as fnc_passOut is where the update is broadcast, and passout is always called.

costno committed:
Update f/medical/fam_bandage.sqf

Does not, in fact, need to be two actions

Co-authored-by: shadow-fa <shadow-fa@users.noreply.github.com>

costno commented on f/medical/fam_bandage.sqf:
```diff
@@ -0,0 +1,165 @@
+	}; 
+
+	// Let the wounded know someone is trying to save them. 
+	if !(_target getVariable ["FAM_CONSCIOUS",true]) then {[["Someone is helping you", "PLAIN"]] remoteExec ["titleText",_target];}; // TODO Test?
```
I haven't gone down much during testing because I'm too good at this game but I'm told that this is working from fairly reliable sources.

costno commented on f/medical/fn_famPassOut.sqf:
```diff
@@ -0,0 +1,117 @@
+if (vehicle _unit != _unit) then {
+
+    if (driver vehicle _unit == _unit) then {
+        _anim = getArray (configfile >> "CfgMovesMaleSdr" >> "States" >> animationState _unit >> "interpolateTo");
+        _unit playMove (_anim select 0);
```
This suggestion won't quite work as _anim = ""; will teleport you above the vehicle standing up if you end up with an interpolateTo that is not defined.  

I am curious exactly when this breaks though without your change as this method was used in SWS and I don't recall seeing issues specifically related to that then?

shadow-fa replied:
Okay for CfgMovesMaleSdr all states have that property.

testing code for reference:
```sqf
{
  if(!isArray (configFile >> "CfgMovesMaleSdr" >> "States" >> _x >> "interpolateTo")) then {
  systemChat str _x;
} forEach ((configFile >> "CfgMovesMaleSdr" >> "States") call BIS_fnc_getCfgSubClasses);
```

shadow-fa replied:
Useful reference: https://github.com/CBATeam/CBA_A3/blob/master/addons/common/fnc_getUnitDeathAnim.sqf

costno committed:
JIP Functionality AFAIK

costno committed:
Merge branch 'medical' of https://github.com/folkarps/F3 into medical

costno committed:
Vehicle animations and medic modifier

Possible resolution for all notes from previous testing

costno committed:
Vehicle gear changes

costno committed:
Mission testing v26 passing

NikkoJT commented:
As Shado' mentioned before, I'd like to have the `FAM_` variable names converted to the `f_var_` standard used elsewhere in FA3

costno committed:
Changes

costno committed:
A few months of tweaks.

costno committed:
Now with UI!

NikkoJT committed:
Fix missing UI config parts

NikkoJT committed:
Comments

NikkoJT committed:
Dead colour distinct from lightly injured colour

NikkoJT committed:
hehe

NikkoJT committed:
remoteExec is already scheduled

NikkoJT committed:
Varnames and comments

NikkoJT committed:
Consolidate remoteExec

NikkoJT committed:
Colour adjustment

NikkoJT committed:
Improved colour handling

NikkoJT committed:
fix bandage action

NikkoJT committed:
fix more var names

NikkoJT committed:
Update dead colour handling

NikkoJT committed:
Remove unneeded bleed text

NikkoJT committed:
less stupid bleed text fix

costno committed:
Prevent pass out message if AI team member is downed.

costno committed:
Clear any message popup when dead.

NikkoJT committed:
spelling

costno committed:
wake up fak reset

costno committed:
Speculative balance changes

Also making sure your downed variable is set before any other scripts run in pass out.

NikkoJT committed:
Remove old CLS

costno committed:
Someone with a math degree looked at the damage handler

And some other stuff

costno committed:
attachTo not enableSimulation

NikkoJT committed:
improves

NikkoJT committed:
hehe

costno committed:
initPlayerLocal works in JIP

Plus some math adjustments and removing excess code.

costno committed:
Merge branch 'medical' of https://github.com/folkarps/F3 into medical

costno committed:
MedSwap for Zeus entities

costno commented:
What's our todo list at this point? 

- Try to make init.sqf work for JIP
- Find and fix any remaining major bugs (my current list is no obvious bugs?)
- Verify equipment quantities proposals are suitable/reasonable
- Verify we're happy with AI skill levels and damage handler levels
- Support for CDLC damage handler levels(?)
- Nikko wants to be more specific with ragdoll force direction

NikkoJT committed:
Adjust ragdoll force based on impact

NikkoJT committed:
Fix direction of force

costno committed:
Passout sounds

costno committed:
Better medical vehicle loads

And I guess bringing vehicle loadouts to 3.5.7

costno committed:
UAV Terminals were not removed while downed

costno committed:
Was showing light injuries but could not treat in damage less than 0.1

costno committed:
Correct JIP operations when called from init.sqf

costno committed:
oops

NikkoJT committed:
Unique JIP ID to prevent overwrites

costno committed:
Damage handler tweak and drag fix

costno committed:
new skill

costno committed:
Separate out legs and hands during bleed/heal ticks

costno committed:
GPS/Map Removal

costno committed:
Balance tweaks

Slightly slowly trickle heal rate. Increased force down threshold

costno committed:
Respawn support?

I think this works

costno committed:
assignGear cleanup

Syncing with other PRs to make the merge easier hopefully

costno committed:
Final final v2 forcedown tweaks

SamLex commented on init.sqf:
```diff
@@ -218,3 +211,20 @@ f_var_viewDistance_crewOnly = true;
 
 // ====================================================================================
+
+if (isServer) then {
```
I feel like this should maybe be built into e.g. famInit rather than being a big blob of code at the bottom of init.sqf. If we want this to be optional (MMs can turn off the swap behaviour), I'd instead suggest a boolean parameter to famInit that an MM could flip instead

SamLex commented on f/spect/fn_activateSpectator.sqf:
```diff
@@ -2,18 +2,22 @@ if (f_param_debugMode == 1) then {
 
-// 'Cinematic' delay before spectator activates
-sleep 3;
+0 spawn {
```
Out of interest, why is this block now scheduled (/spawned)?

costno replied:
unclear! I'll doublecheck that it isn't necessary

costno replied:
Aha, somewhere along the way and only in my mission test file I had a isCall added to the onPlayerRespawn template in description.ext

SamLex commented on f/medical/fn_FAMdiagnose.sqf:
This file seems to have a lot of commented out code. Can this be removed?

SamLex commented on f/medical/fn_famAddBandageAction.sqf:
Same in this file re: commented out code

SamLex commented on f/medical/fn_famAddDiagnoseAction.sqf:
```diff
@@ -0,0 +1,137 @@
+
+// INITIALIZE
+params ["_unit"];
+/*
```
This whole block is commented out?

SamLex commented on f/medical/fn_famAddHealAction.sqf:
```diff
@@ -0,0 +1,133 @@
+private _healProg = "(_target distance _caller < 3) && {alive _target && !(_target getVariable ['f_var_fam_conscious',true])}"; // This one is always the same, start condition varies by unit type.
+private _healTime = 6; // Action Duration
+private _healMedicTime = 4.5; // Action Duration
+/*
```
Commented out code

SamLex commented on f/medical/fn_famDamageHandler.sqf:
```diff
@@ -0,0 +1,86 @@
+// debug
+if (f_param_debugMode == 1) then
+{
+	systemChat format ["(fnc_famDamageHandler): %1 | %2 | %3 | %4 | %5",_unit,_selection,_currentDamage /*toFixed 4*/,_damage /*toFixed 4*/,_newDamage /*toFixed 4*/];
```
Its debug logs so less important, but can the `toFixed` comments be removed?

SamLex reviewed (commented):
A few small comments on the periphery code. There also seems to be quite a lot of commented out code that would be good to clean-up to prevent future confusion.

I only did a quick scan of the bulk of the Medical system as it's has enough iteration & testing that I don't think it needs a deep review (and I don't want to suggest any changes at this point)

costno committed:
Cleanup

Dead comments and weird scheduling

costno committed:
oops v2

SamLex committed:
Merge remote-tracking branch 'origin/dev' into medical
